### PR TITLE
Spmi third cleaning

### PR DESCRIPTION
**Pull request by sandreenko**
1. delete duplicate code for sigInfo;
2. fix the names for pSig that should be pSig_Index when it is the index of the buffer;
3. create dump class, that returns string representation of most popular structs, allows to delete duplicates. 
4. dump helper also shows some forgotten fields in dmp functions, fixed.

### CODE REVIEW

**Comment by sandreenko** [comment]
This PR allows me to support R2R methods without new duplicates.

**Comment by BruceForstall** [comment]
@dotnet/jit-contrib

**Comment by BruceForstall** [comment]
You'll need to add spmidumphelper.cpp to the internal VS project files as well.

File: src/ToolBox/superpmi/superpmi-shared/spmidumphelper.cpp [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+#include "standardpch.h"
+#include "spmidumphelper.h"
+
+std::string SpmiDumpHelper::DumpAgnostic_CORINFO_RESOLVED_TOKENin(
+    const MethodContext::Agnostic_CORINFO_RESOLVED_TOKENin& tokenIn)
+{
+    char buffer[MAX_BUFFER_SIZE];
+    sprintf(buffer, "tc - % 016llX ts - % 016llX tok - %08X tt - %u", tokenIn.tokenContext, tokenIn.tokenScope,
+            tokenIn.token, tokenIn.tokenType);
```
Comment (BruceForstall) [inline src/ToolBox/superpmi/superpmi-shared/spmidumphelper.cpp]: Does it work to have a space between `%` and `016llX`? IMO, you should remove the space.

Comment (BruceForstall) [inline src/ToolBox/superpmi/superpmi-shared/spmidumphelper.cpp]: I'd also, for consistency, remove the spaces surrounding the `-`

---
In reply to: [123583653](https://github.com/dotnet/coreclr/pull/12418#discussion_r123583653) [](ancestors = 123583653)

File: src/ToolBox/superpmi/superpmi-shared/methodcontext.cpp [hunk 2]
```diff
@@ -3349,7 +3199,7 @@ void MethodContext::dmpGetFieldInfo(const Agnostic_GetFieldInfo& key, const Agno
                 break;
         }
     }
-    printf("}");
+    printf("}\n");
 }
```
Comment (BruceForstall) [inline src/ToolBox/superpmi/superpmi-shared/methodcontext.cpp]: I think the dmp functions don't add trailing newlines; the caller does.

File: src/ToolBox/superpmi/superpmi-shared/methodcontext.cpp [hunk 3]
```diff
@@ -3661,19 +3511,14 @@ void MethodContext::recCanAccessClass(CORINFO_RESOLVED_TOKEN*      pResolvedToke
...
+    printf("CanAccessClass key rt{%s}, callerHandle %016llX\n",
+           SpmiDumpHelper::DumpAgnostic_CORINFO_RESOLVED_TOKEN(key.ResolvedToken).c_str(), key.callerHandle);
     printf(", value hnum-%u na-%u {", value.AccessHelper.helperNum, value.AccessHelper.numArgs);
     for (int i = 0; i < CORINFO_ACCESS_ALLOWED_MAX_ARGS; i++)
     {
         printf("{%016llX %u}", value.AccessHelper.args[i].constant, value.AccessHelper.args[i].argType);
     }
-    printf("} res-%u", value.result);
+    printf("} res-%u\n", value.result);
 }
```
Comment (BruceForstall) [inline src/ToolBox/superpmi/superpmi-shared/methodcontext.cpp]: no trailing newline?

File: src/ToolBox/superpmi/superpmi-shared/methodcontext.cpp (hunk 2 above)
Comment (sandreenko) [inline src/ToolBox/superpmi/superpmi-shared/methodcontext.cpp]: You are right. Fixed, also deleted trailing newlines from dmpGetInlinedCallFrameVptr.

**Pull request merged**